From c8e56913ff414fe18d8ba5675819c7eaef4946cb Mon Sep 17 00:00:00 2001
From: Shawn Rutledge <shawn.rutledge@qt.io>
Date: Wed, 23 Sep 2026 22:23:16 +0200
Subject: [PATCH] Update md4c to 0.6.0
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Amends d377d1f3a921488525f6e75c850632f559707747 : Qt's definition of
GitHub dialect remains as it was, while md4c has recently added features
that we don't yet support. We need to verify that the new features are
not inadvertently turned on before Qt is ready to handle them, and deal
with the clashing feature flag (0x100000 is for FrontMatter in Qt,
footnotes in md4c). It seems for now that we will not be able to support
footnotes, both because of the clashing flag and because it would be
hard to change the text-editing components to support them; so a test is
added to verify that the markdown footnote syntax is NOT interpreted
specially. Likewise, a test verifies that a markdown "admonition" is
treated as an ordinary block quote for now (but this can be improved
later).

[ChangeLog][Third-Party Code] Updated md4c to version 0.6.0.

Fixes: QTBUG-148738
Fixes: QTBUG-150636
Fixes: QTBUG-150661
Task-number: QTBUG-149921
Pick-to: 6.12.0 6.8 6.5
Change-Id: Iffd17982fe0d606b0f5003e25ee85b2db6ad019a
Reviewed-by: Oliver Eftevaag <oliver.eftevaag@qt.io>
Reviewed-by: Kai Köhne <kai.koehne@qt.io>
(cherry picked from commit 032643f99e090e0304ef935c2ee11c109d55c8a1)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
diff --git a/src/gui/text/qtextmarkdownimporter.cpp b/src/gui/text/qtextmarkdownimporter.cpp
index 3f57c700fa83..3ee8b4104dee 100644
--- a/src/gui/text/qtextmarkdownimporter.cpp
+++ b/src/gui/text/qtextmarkdownimporter.cpp
@@ -51,8 +51,10 @@ static_assert(int(QTextMarkdownImporter::FeaturePermissiveAutoLinks) == MD_FLAG_
 static_assert(int(QTextMarkdownImporter::FeatureTasklists) == MD_FLAG_TASKLISTS);
 static_assert(int(QTextMarkdownImporter::FeatureNoHTML) == MD_FLAG_NOHTML);
 static_assert(int(QTextMarkdownImporter::DialectCommonMark) == MD_DIALECT_COMMONMARK);
+// We don't yet support all the features in MD_DIALECT_GITHUB
 static_assert(int(QTextMarkdownImporter::DialectGitHub) ==
-              (MD_DIALECT_GITHUB | MD_FLAG_UNDERLINE | QTextMarkdownImporter::FeatureFrontMatter));
+              (MD_FLAG_PERMISSIVEAUTOLINKS | MD_FLAG_TABLES | MD_FLAG_STRIKETHROUGH | MD_FLAG_TASKLISTS
+                | MD_FLAG_UNDERLINE | QTextMarkdownImporter::FeatureFrontMatter));
 
 // --------------------------------------------------------
 // MD4C callback function wrappers
@@ -167,7 +169,7 @@ void QTextMarkdownImporter::import(const QString &markdown)
 {
     MD_PARSER callbacks = {
         0, // abi_version
-        unsigned(m_features),
+        unsigned(m_features) & unsigned(0xFFEFFFFF), // exclude MD_FLAG_FOOTNOTES: unsupported in Qt
         &CbEnterBlock,
         &CbLeaveBlock,
         &CbEnterSpan,
diff --git a/src/gui/text/qtextmarkdownimporter_p.h b/src/gui/text/qtextmarkdownimporter_p.h
index 8b8f4ec9bb80..ba188a3b33e8 100644
--- a/src/gui/text/qtextmarkdownimporter_p.h
+++ b/src/gui/text/qtextmarkdownimporter_p.h
@@ -47,6 +47,10 @@ class Q_GUI_EXPORT QTextMarkdownImporter
         FeatureTasklists =                0x0800,
         FeatureUnderline =                0x4000,
         FeatureFrontMatter =            0x100000, // Qt feature, not yet in MD4C
+        // NOTE: flags above are kept in sync with #define MD_FLAG_*; therefore,
+        // do NOT add any more Qt-specific flags (within the lower 32 bits).
+        // Because we have FeatureFrontMatter, we cannot support MD_FLAG_FOOTNOTES.
+
         // composite flags
         FeaturePermissiveAutoLinks = FeaturePermissiveMailAutoLinks
             | FeaturePermissiveURLAutoLinks | FeaturePermissiveWWWAutoLinks,
